### PR TITLE
Block requests to join ring after failed bootstraps

### PR DESCRIPTION
**Pull request by zpear**
Had to mock reads from System keyspace b/c it was messing with some unit tests

zombie mode should be fine b/c it goes through join ring 

Validated enable interfaces endpoint throws. Also circumvented w/ nodetool and validated we don't set tokens as part of enabling gossip either:

```
./service/bin/nodetool enablegossip
error: Cannot set tokens for a non-bootstrapped node
-- StackTrace --
com.palantir.cassandra.db.BootstrappingSafetyException: Cannot set tokens for a non-bootstrapped node
        at org.apache.cassandra.service.StorageService.setGossipTokens(StorageService.java:252)
```

### CODE REVIEW

File: src/java/org/apache/cassandra/service/StorageService.java [hunk 1]
```diff
@@ -641,6 +650,7 @@ public synchronized void initServer(int delay) throws ConfigurationException
         logger.info("Thrift API version: {}", cassandraConstants.VERSION);
         logger.info("CQL supported versions: {} (default: {})",
                     StringUtils.join(ClientState.getCQLSupportedVersion(), ","), ClientState.DEFAULT_CQL_VERSION);
+        isBootstrapMode = SystemKeyspace.getBootstrapState() == SystemKeyspace.BootstrapState.IN_PROGRESS;
```
Comment (zpear) [inline src/java/org/apache/cassandra/service/StorageService.java]: this is one of the critical lines: some assertions based on isBootstrapMode, like read verb handler: https://github.com/palantir/cassandra/blob/2bcaf4adf07309a1cc9c11da03f44136204d6cf6/src/java/org/apache/cassandra/db/ReadVerbHandler.java#L37 may not reject reads otherwise, as `isBootstrapMode` would be null instead of `true` on second startup

File: src/java/org/apache/cassandra/service/StorageService.java [hunk 2]
```diff
@@ -247,6 +247,10 @@ public void setTokens(Collection<Token> tokens)
 
     public void setGossipTokens(Collection<Token> tokens)
     {
+        if (DatabaseDescriptor.isAutoBootstrap() && !bootstrapComplete())
```
Comment (inespot) [inline src/java/org/apache/cassandra/service/StorageService.java:250]: Do we ever expect auto_bootstrap to be set to false?

File: src/java/org/apache/cassandra/service/StorageService.java (hunk 1 above)
Comment (inespot) [reply src/java/org/apache/cassandra/service/StorageService.java]: Nit: there is an existing function for this check: `SystemKeyspace.bootstrapInProgress()`

File: test/unit/org/apache/cassandra/service/ActiveRepairServiceTest.java [hunk 3]
```diff
@@ -90,6 +87,7 @@ public void prepare() throws Exception
 
         TokenMetadata tmd = StorageService.instance.getTokenMetadata();
         tmd.clearUnsafe();
+        SystemKeyspace.setBootstrapState(SystemKeyspace.BootstrapState.COMPLETED);
```
Comment (inespot) [inline test/unit/org/apache/cassandra/service/ActiveRepairServiceTest.java:90]: Why is this change needed for this and the MutationsVerificationUtils tests?

File: src/java/org/apache/cassandra/service/StorageService.java (hunk 2 above)
Comment (zpear) [reply src/java/org/apache/cassandra/service/StorageService.java:250]: we never set it, but I included here to be consistent w/ the rest of the codebase

File: test/unit/org/apache/cassandra/service/ActiveRepairServiceTest.java (hunk 3 above)
Comment (zpear) [reply test/unit/org/apache/cassandra/service/ActiveRepairServiceTest.java:90]: I believe the test initialization requires it here and the other test suite w/ the new assertions

Comment (inespot) [reply test/unit/org/apache/cassandra/service/ActiveRepairServiceTest.java:90]: Because `setTokens` calls `setGossipTokens`, got it

**Review by inespot — APPROVED** [review]
